### PR TITLE
Add Token retrieval and search for claim

### PR DESCRIPTION
**Pull request by corinnekrych**
Ready for review

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -89,23 +72,74 @@ $ cd $GOPATH/src/github.com/fabric8-services/fabric8-toggles-service
...
+
+image::images/debug.png[GoLand Debug]
+====
+
+`Configure unleash database`::
+
+* Go to fabric8-toggles-server admin console (ie: the unleash admin console)
+   http://fabric8-toggles-f8toggles.MINISHIFT_IP.nip.io/#/strategies add a strategy with name `enableByGroupID` with a parameter `groupID`
+* Go to features list, add a feature with name "feature1", give a description and add the newly created strategy, enter `Red Hat`
+for `groupId`. Enable the feature, you should get:
```
Comment (xcoulon) [inline README.adoc]: groupID

File: controller/features.go [hunk 2]
```diff
@@ -2,29 +2,103 @@ package controller
...
+	"github.com/fabric8-services/fabric8-toggles-service/featuretoggles"
+	"github.com/fabric8-services/fabric8-wit/log"
 	"github.com/goadesign/goa"
+	goajwt "github.com/goadesign/goa/middleware/security/jwt"
 	uuid "github.com/satori/go.uuid"
+	"strings"
 )
 
 // FeaturesController implements the features resource.
 type FeaturesController struct {
```
Comment (xcoulon) [inline controller/features.go]: we should probably use a single controller for all requests related to features, but that can be changed in another PR

File: controller/features.go [hunk 3]
```diff
@@ -2,29 +2,103 @@ package controller
...
 
-	// Put your logic here
+	jwtToken := goajwt.ContextJWT(ctx)
+	groupId := ""
+	if jwtToken != nil {
+		groupId = jwtToken.Claims.(jwtgo.MapClaims)["company"].(string) // TODO replace with a custom claim group_id
+	}
+	var enableFeatures *app.FeatureList
+	if c.config != nil {
+		featuresListFromUnleash, err := getFeatureListFromUnleashServer(ctx, c.config, groupId)
```
Comment (xcoulon) [inline controller/features.go]: for your tests, you could define an `unleashService` that has its own `unleashClient`, then replace that client with a mock, so you can do whatever you want without depending on an actual deployed unleash server

File: controller/features.go [hunk 4]
```diff
@@ -2,29 +2,103 @@ package controller
...
+		if err != nil {
+			log.Error(ctx.Context, map[string]interface{}{
+				"err": err,
+			}, "Unable to connect to Unleash server")
+			return ctx.Err()
+		}
+		log.Debug(ctx, nil, "FEATURES: %s", featuresListFromUnleash)
+		enableFeatures = featuresListFromUnleash
+	} else { // TEST Data
```
Comment (xcoulon) [inline controller/features.go]: see comment above, so you can remove that part.

File: controller/features.go [hunk 5]
```diff
@@ -2,29 +2,103 @@ package controller
...
+		}, "Unable to connect to Unleash server")
+		return nil, err
+	}
+
+	listOfFeatures := toggleClient.GetEnabledFeatures(groupId)
+	res = format(listOfFeatures, groupId)
+	return &res, nil
+}
+
+func format(list []string, groupId string) app.FeatureList {
```
Comment (xcoulon) [inline controller/features.go]: we usually name this kind of function `convert`

File: controller/features.go [hunk 6]
```diff
@@ -2,29 +2,103 @@ package controller
...
+		ID := uuid.NewV4()
+		descriptionFeature := "Description of the feature"
+		enabledFeature := true
+		nameFeature := list[i]
+
+		feature := app.Feature{
+			ID: ID,
+			Attributes: &app.FeatureAttributes{
+				Description: &descriptionFeature,
+				Enabled:     &enabledFeature,
```
Comment (xcoulon) [inline controller/features.go]: why would you need that `Enabled` attribute ?

File: controller/features.go [hunk 7]
```diff
@@ -2,29 +2,103 @@ package controller
...
+		// TODO remove ID, make unleash client return description
+		ID := uuid.NewV4()
+		descriptionFeature := "Description of the feature"
+		enabledFeature := true
+		nameFeature := list[i]
+
+		feature := app.Feature{
+			ID: ID,
+			Attributes: &app.FeatureAttributes{
+				Description: &descriptionFeature,
```
Comment (xcoulon) [inline controller/features.go]: can you use a non-hardcoded value here ?

File: controller/features.go [hunk 8]
```diff
@@ -2,29 +2,103 @@ package controller
...
+				Name:        &nameFeature,
+				GroupID:     &groupId,
+			},
+		}
+		res.Data = append(res.Data, &feature)
+	}
+	return res
 }
 
 func buildFeaturesList(length int) *app.FeatureList {
```
Comment (xcoulon) [inline controller/features.go]: I assume that this function can be removed (see comment above)

File: controller/features_test.go [hunk 9]
```diff
@@ -2,21 +2,29 @@ package controller
...
 	"context"
+	jwt "github.com/dgrijalva/jwt-go"
 	"github.com/fabric8-services/fabric8-toggles-service/app/test"
 	"github.com/goadesign/goa"
+	goajwt "github.com/goadesign/goa/middleware/security/jwt"
 	"github.com/magiconair/properties/assert"
 	"testing"
 )
 
 func TestFeatures(t *testing.T) {
```
Comment (xcoulon) [inline controller/features_test.go]: I would rename this function to `TestListFeatures` so we know you're testing the `List` operation

File: controller/features_test.go [hunk 10]
```diff
@@ -2,21 +2,29 @@ package controller
...
 	svc := goa.New("feature")
-	ctrl := NewFeaturesController(svc)
+	ctrl := NewFeaturesController(svc, nil)
 	expectedFeaturesList := buildFeaturesList(5)
 
-	t.Run("OK", func(t *testing.T) {
+	t.Run("OK with no token", func(t *testing.T) {
 		_, featuresList := test.ListFeaturesOK(t, context.Background(), svc, ctrl)
```
Comment (xcoulon) [inline controller/features_test.go]: aren't you supposed to get an auth error here ?

File: controller/features_test.go [hunk 11]
```diff
@@ -2,21 +2,29 @@ package controller
...
-		assert.Equal(t, 5, len(featuresList.Data))
-		assert.Equal(t, featuresList.Data[0].Attributes.Name, expectedFeaturesList.Data[0].Attributes.Name)
+		assert.Equal(t, 0, len(featuresList.Data))
+		//assert.Equal(t, featuresList.Data[0].Attributes.Name, expectedFeaturesList.Data[0].Attributes.Name)
+	})
+	t.Run("OK with jwt token", func(t *testing.T) {
+		_, featuresList := test.ListFeaturesOK(t, createValidContext(), svc, ctrl)
+		assert.Equal(t, 2, len(featuresList.Data))
```
Comment (xcoulon) [inline controller/features_test.go]: you should use `require` instead of `assert`, so the execution will stop if `featureList` is nil or does not have the expected size. Otherwise, the execution will continue and might result in a `panic`

File: controller/features_test.go [hunk 12]
```diff
@@ -25,3 +33,9 @@ func TestFeatures(t *testing.T) {
 	//	test.ListFeaturesNotFound(t, context.Background(), svc, ctrl)
 	//})
 }
+func createValidContext() context.Context {
```
Comment (xcoulon) [inline controller/features_test.go]: so you should have an `invalid` context, too, in order to verify that you get no enabled feature for a non-Red Hat user

File: featuretoggles/unleash_client.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+// Name the name of the stragegy. Must match the name on the Unleash server.
+func (s *EnableByGroupIDStrategy) Name() string {
+	return EnableByGroupID
+}
+
+// IsEnabled returns `true` if the given context is compatible with the settings configured on the Unleash server
+func (s *EnableByGroupIDStrategy) IsEnabled(settings map[string]interface{}, ctx *unleashcontext.Context) bool {
+	fmt.Printf("Checking %+v vs %+v", settings, ctx)
```
Comment (xcoulon) [inline featuretoggles/unleash_client.go]: you should replace that with a `log.Debug(..)`

File: controller/features.go (hunk 6 above)
Comment (corinnekrych) [reply controller/features.go]: @xcoulon we should change: to return a list of struct bc atm we return only a list of string (with feature name) see: https://github.com/xcoulon/unleash-client-go/commit/377eea70f16806a6f68414ceed39b1f750eafe1f#diff-d0b7f86cea13f90d3cc0cc2338da8d5fR266

**Comment by corinnekrych** [comment]
closing this one in favor of #5
